File: frontend/src/api/spool.ts
import client from './client'

export interface SpoolJobInfo {
  jobid: string
  jobname: string
  owner: string
  status: string
  submitted: string
  queue: string
  prty: number
  job_class: string
}

export interface SpoolJobDetail extends SpoolJobInfo {
  output: string
}

export async function listSpoolJobs(): Promise<SpoolJobInfo[]> {
  const res = await client.get('/spool/jobs')
  return res.data.data as SpoolJobInfo[]
}

export async function getSpoolJob(jobid: string): Promise<SpoolJobDetail> {
  const res = await client.get(`/spool/jobs/${jobid}`)
  return res.data.data as SpoolJobDetail
}

export async function submitJcl(jcl: string, owner: string): Promise<SpoolJobInfo> {
  const res = await client.post('/spool/submit', { jcl, owner })
  return res.data.data as SpoolJobInfo
}

export async function deleteSpoolJob(jobid: string): Promise<void> {
  await client.delete(`/spool/jobs/${jobid}`)
}

export async function cancelSpoolJob(jobid: string): Promise<void> {
  await client.post(`/spool/jobs/${jobid}/cancel`)
}

File: backend/app/routers/spool.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel, validator
from app.dependencies import get_jobs, JobEngine
from app.models.responses import ok, err, APIResponse

router = APIRouter(prefix="/api/spool", tags=["spool"])

_JCL_MAX_BYTES = 64_000


class SubmitBody(BaseModel):
    jcl: str
    owner: str = "TOMTZ"

    @validator('jcl')
    def jcl_max_size(cls, v: str) -> str:
        if len(v.encode()) > _JCL_MAX_BYTES:
            raise ValueError("JCL exceeds maximum size of 64KB")
        return v


def _job_meta(job):
    return {
        "jobid": job.jobid,
        "jobname": job.jobname,
        "owner": job.owner,
        "status": job.status,
        "submitted": job.submitted,
        "queue": job.queue,
        "prty": job.prty,
        "job_class": job.job_class,
    }


@router.get("/jobs", response_model=APIResponse)
def list_jobs(jobs: JobEngine = Depends(get_jobs)):
    return ok([_job_meta(j) for j in jobs.list_jobs()])


@router.get("/jobs/{jobid}", response_model=APIResponse)
def get_job(jobid: str, jobs: JobEngine = Depends(get_jobs)):
    job = jobs.get_job(jobid)
    if job is None:
        return err(f"Job {jobid} not found in spool")
    return ok({**_job_meta(job), "output": job.output, "jcl": job.jcl})


@router.delete("/jobs/{jobid}", response_model=APIResponse)
def delete_job(jobid: str, jobs: JobEngine = Depends(get_jobs)):
    if jobs.delete_job(jobid):
        return ok({"deleted": jobid})
    return err(f"Job {jobid} not found in spool")


@router.post("/jobs/{jobid}/cancel", response_model=APIResponse)
def cancel_job(jobid: str, jobs: JobEngine = Depends(get_jobs)):
    if jobs.cancel_job(jobid):
        return ok({"cancelled": jobid})
    return err(f"Job {jobid} not found in spool")


@router.post("/submit", response_model=APIResponse)
def submit_jcl(body: SubmitBody, jobs: JobEngine = Depends(get_jobs)):
    try:
        job = jobs.submit(body.jcl, body.owner)
        return ok(_job_meta(job))
    except ValueError as e:
        return err(str(e))
